amethyst_ui/src/selection_order_cache.rs

```diff
@@ -0,0 +1,128 @@
+use amethyst_core::specs::Entities;
+use amethyst_core::specs::Entity;
+use amethyst_core::specs::Join;
+use amethyst_core::specs::System;
+use amethyst_core::specs::Write;
+use amethyst_core::specs::WriteStorage;
+use hibitset::BitSet;
+use std::cmp::Ordering;
+use std::marker::PhantomData;
+
+use amethyst_core::specs::ReadStorage;
+
+use {Selectable, Selected};
+
+/// A cache sorted by tab order and then by Entity.
+/// Used to quickly find the next or previous selectable entities.
+#[derive(Debug, Clone, Default)]
+pub struct CachedSelectionOrder {
+    /// The cached bitset.
+    pub cached: BitSet,
+    /// The cache holding the z position and the corresponding entity.
+    pub cache: Vec<(u32, Entity)>,
+}
+
+impl CachedSelectionOrder {
+    // TODO: Change WriteStorage to ReadStorage. @torkleyy
+    //
+    //
+    //
+    //
+    //
+    /// Returns the index of the highest cached element (index in the cache!) that is currently selected.
+    pub fn highest_order_selected_index(
+        &self,
+        selected_storage: &WriteStorage<Selected>,
+    ) -> Option<usize> {
+        self.cache
```

Comment:
Depending on how many times you would theoretically read from this and how many elements will be in here, it might make more sense to calculate this value once on insert and keep it around on a field on the struct. You could even keep track of the next highest tab order in here and then skip the calculation if you push a new element that is below it.

Reply:
I'll add this to the todo list. It is a very good idea. Thanks!